services/blockstorage/test/block_sync_server_test.go

```diff
@@ -0,0 +1,129 @@
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestSourceRespondToAvailabilityRequests(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		harness := newHarness(ctx)
+		harness.setupSomeBlocks(3)
+		harness.gossip.When("SendBlockAvailabilityResponse", mock.Any).Return(nil, nil).Times(1)
+		msg := builders.BlockAvailabilityRequestInput().
+			WithFirstBlockHeight(1).
+			WithLastCommittedBlockHeight(primitives.BlockHeight(2)).
+			WithLastBlockHeight(primitives.BlockHeight(2)).
+			Build()
+		_, err := harness.blockStorage.HandleBlockAvailabilityRequest(msg)
+
+		require.NoError(t, err, "expecting a happy flow")
+
+		harness.verifyMocks(t, 1)
+	})
+}
+
+func TestSourceRespondsNothingToAvailabilityRequestIfSourceIsBehindPetitioner(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+
+		harness := newHarness(ctx)
+		harness.expectCommitStateDiff()
+		harness.commitBlock(builders.BlockPair().WithHeight(primitives.BlockHeight(1)).Build())
+
+		harness.gossip.Never("SendBlockAvailabilityResponse", mock.Any)
+
+		msg := builders.BlockAvailabilityRequestInput().WithLastCommittedBlockHeight(primitives.BlockHeight(20)).Build()
+		_, err := harness.blockStorage.HandleBlockAvailabilityRequest(msg)
+
+		require.NoError(t, err, "expecting a happy flow")
+		harness.verifyMocks(t, 1)
+	})
+}
+
+func TestSourceIgnoresSendBlockAvailabilityRequestsIfFailedToRespond(t *testing.T) {
+
+	test.WithContext(func(ctx context.Context) {
+		harness := newHarness(ctx)
+		harness.setupSomeBlocks(3)
+
+		harness.gossip.When("SendBlockAvailabilityResponse", mock.Any).Return(nil, errors.New("gossip failure")).Times(1)
+		msg := builders.BlockAvailabilityRequestInput().
+			WithFirstBlockHeight(1).
+			WithLastCommittedBlockHeight(primitives.BlockHeight(2)).
+			WithLastBlockHeight(primitives.BlockHeight(2)).
+			Build()
+		_, err := harness.blockStorage.HandleBlockAvailabilityRequest(msg)
+
+		require.Error(t, err, "expecting an error from the server event flow")
+
+		harness.verifyMocks(t, 1)
+	})
+}
+
+func TestSourceRespondsWithChunks(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		batchSize := uint32(10)
+		harness := newHarness(ctx).withBatchSize(batchSize).withNodeKey(keys.Ed25519KeyPairForTests(4).PublicKey())
+		lastBlock := 12
+		harness.setupSomeBlocks(lastBlock)
+
+		firstHeight := primitives.BlockHeight(1)
+		lastHeight := primitives.BlockHeight(10) // hardcoding this, but it is a function of the batchSize
+
+		msg := builders.BlockSyncRequestInput().
+			WithSenderPublicKey(keys.Ed25519KeyPairForTests(1).PublicKey()).
+			WithFirstBlockHeight(firstHeight).
+			Build()
+
+		chunksResponseVerifier := func(i interface{}) bool {
+			response, ok := i.(*gossiptopics.BlockSyncResponseInput)
+			if !ok {
+				require.Failf(t, "response type does not match", "", i)
+			}
+			require.EqualValues(t, batchSize, len(response.Message.BlockPairs), "batch size does not match config")
```

Comment:
`require.Len`?

Reply:
need to get myself used to look at require.X